### ISSUE TITLE
'rspamc learn_spam' fails with 'HTTP error : 500, Unknown statistics error'

### ISSUE THREAD

**Issue by yasuhirokimura** [rspamd/rspamd#2678]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [x] Serious bug
* [ ] Ordinary bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:

1.8.3

### Operation system, CPU:

FreeBSD 11.2-RELEASE amd64

### Description (Please provide a descriptive summary of the issue):

About 2 weeks ago I found `rspamc learn_spam` fails as following.

yasu@eastasia[2041]% rspamc learn_spam ~/tmp/spam01.txt                                                                    ~
Results for file: /home/yasu/tmp/spam01.txt (0.024 seconds)
HTTP error: 500, Unknown statistics error

I can't say exactly when but in the past it certainly worked well.

I removed all files under ${DBDIR} and start spam/ham learning from
scratch. At first 'rspamc learn_spam' completed successfully. But
after a week or so same error happened again.

`rspamc learn_ham` always succeeds.

### Relevant logs (see details [here](https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd)):

When executing `rspamc learn_spam` following messages are written to
log file.

```
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_controller_check_password: allow unauthorized connection from a trusted IP 127.0.0.1
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_message_parse: loaded message; id: <20181127235149.DF964AE5EA9@mail.prospector.kz>; queue-id: <undef>; size: 4210; checksum: <21b55fd227635acf41c417c420722c42>                          
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_mime_text_part_utf8_convert: converted from ISO_8859-1:1987 to UTF-8 inlen: 811, outlen: 811
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_mime_part_detect_language: detected part language: en       
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_stat_classifiers_process: skip statistics as SPAM class is missing
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_task_process: learn error: Unknown statistics error         
2018-12-01 19:54:07 #752(controller) <731c60>; csession; rspamd_controller_learn_fin_task: cannot learn <20181127235149.DF964AE5EA9@mail.prospector.kz>: Unknown statistics error
```

### Expected results:

`rspamc learn_spam` completes successfully.

### Actual results:

`rspamc learn_spam` fails as following.

yasu@eastasia[2041]% rspamc learn_spam ~/tmp/spam01.txt                                                                    ~
Results for file: /home/yasu/tmp/spam01.txt (0.024 seconds)
HTTP error: 500, Unknown statistics error

### Configuration (e.g. `rspamadm configdump module`):

I don't make any customization about spam learning.

**Comment by vstakhov** [comment]
IDK: enable bayes debugging at least.

**Comment by vstakhov** [comment]
Are you using sqlite backend?

**Comment by yasuhirokimura** [comment]
Yes, I think so.

```bash
yasu@eastasia[4693]% rspamadm configdump classifier.bayes.backend
*** Section classifier.bayes.backend ***
"sqlite3"
*** End of section classifier.bayes.backend ***
yasu@eastasia[4694]%
```

**Comment by sem-hub** [comment]
The same problem on FreeBSD with the log:
2019-05-28 13:57:17 #42580(controller) <9c769f>; csession; rspamd_task_process: learn error: Unknown statistics error, found on stage learn; classifier: (null)
2019-05-28 13:57:17 #42580(controller) <9c769f>; csession; rspamd_controller_learn_fin_task: cannot learn <CA+1tZ+6G7fx1eJFe8V9DDTJVH5f1MfyAwz2pKQDY5Tus=JHxOw@mail.gmail.com>: Unknown statistics error, found on stage learn; classifier: (null)

The same backend: sqlite3

**Comment by Twilek-de** [comment]
I am seeing this problem now as well (I don´t know how long it has existed already. I had a redis backend. I have switched to sqlite and now it seems to work again.

**Comment by bong0** [comment]
Since a few weeks I'm having that issue as well, the log seems not so say more than the client output does. (the unknown statistics error)

My backed is redis, default install on Debian using the rspamd.com repo:

```
apt show rspamd
Package: rspamd
Version: 2.0-1~jessie
Maintainer: Vsevolod Stakhov <vsevolod@highsecure.ru>
[...]
APT-Sources: http://rspamd.com/apt-stable/ jessie/main amd64 Packages
```

Debug log:
```
[more ... "token" messages]
2019-10-27 15:28:06 #9075(controller) <966b1d>; bayes; bayes_learn_spam: token 10245623571907066200 <#rcv:45.155.37.0:esmtp:#F:Werner Klemm>: window: 2, total_count: 0, spam_count: 0, ham_count: 0
2019-10-27 15:28:06 #9075(controller) <966b1d>; bayes; bayes_learn_spam: token 539890791086949597 <#rcv:45.155.37.0:esmtp:#h:C-T:text/html; charset="UTF-8">: window: 3, total_count: 0, spam_count: 0, ham_count: 0
2019-10-27 15:28:06 #9075(controller) <966b1d>; bayes; bayes_learn_spam: token 1730648915295258020 <#rcv:45.155.37.0:esmtp:#hh:54e5275>: window: 4, total_count: 0, spam_count: 0, ham_count: 0
2019-10-27 15:28:06 #9075(controller) <966b1d>; csession; rspamd_task_process: learn error: Unknown statistics error, found when storing data on backend; classifier: (null)
2019-10-27 15:28:06 #9075(controller) <966b1d>; csession; rspamd_controller_learn_fin_task: cannot learn <d1fa3087ee3c3bcdb33ae481b9bb5dca@towpartsnow.com>: Unknown statistics error, found when storing data on backend; classifier: (null)
```

**Comment by nahuelange** [comment]
I have the same error

**Comment by snowirbis** [comment]
Same here. FreeBSD 11.3 amd64 rspamd 2.2

**Comment by arnowelzel** [comment]
Same here, Ubuntu 18.04 amd64, rspamd 2.5, but with redis as backend. I get this error about 2 or 3 times per hour:

`learn error: Unknown statistics error, found when storing data on backend; classifier: bayes`

And this is the result of `rspamadm configdump classifier.bayes.backend`:

```
2020-04-10 13:13:38 #0(main) lua_ip_from_string: cannot parse ip:
*** Section classifier.bayes.backend ***
"redis"
*** End of section classifier.bayes.backend ***
```

**Comment by norbusan** [comment]
Just to chime in, same here, too, Debian buster, rspamd 2.5, redis.

**Comment by bong0** [comment]
I finally resolved the issue in my case, though with the now unsupported version 2.2 on debian jessie (8.11). I want to share it nevertheless, it took the following:

1. I removed `/etc/rspamd/local.d/statistic.conf`, which contained 
```
classifier "bayes" {
    servers = "127.0.0.1:6379";
}
```
before.
2. I introduced  `/etc/rspamd/local.d/classifier-bayes.conf` with the contents:
```
servers = "127.0.0.1";
backend = "redis";
```

Then restart rspamd. After uploading ham/spam via rspamc or the webinterface, I started seeing heavy output in `redis-cli monitor` :)

**Comment by R3belWithoutAPause** [comment]
This fixed my problem running Rspamd 2.5 on Slackware. Thank you!

**Comment by v2b1n** [comment]
> I finally resolved the issue in my case, though with the now unsupported version 2.2 on debian jessie (8.11). I want to share it nevertheless, it took the following:
> 
>     1. I removed `/etc/rspamd/local.d/statistic.conf`, which previously contained
> 
> 
> ```
> classifier "bayes" {
>     servers = "127.0.0.1:6379";
> }
> ```
> 
>     1. I introduced  `/etc/rspamd/local.d/classifier-bayes.conf` with the contents:
> 
> 
> ```
> servers = "127.0.0.1";
> backend = "redis";
> ```
> 
>     1. Restart rspamd.
> 
> 
> After uploading ham/spam via rspamc or the webinterface, I started seeing heavy output in `redis-cli monitor` :)

Thanks, fixed rspamd 2.7 @ debian stretch

**Comment by detrout** [comment]
I too had this problem on a system just upgraded to Debian Bullseye using the Debian rspamd 2.7-1, rpamd had previously been working with buster.

Since I noticed /etc/rspamd/statistic.conf already included
`backend = "redis";`

I decided to create /etc/rspamd/local.d/classifier-bayes.conf with just setting the address.

`servers="127.0.0.1";`

After that rspamc worked and stopped giving me the 500 error
